### ISSUE TITLE
Should not warn dereferencing collection after foreach

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#30562]
The second `foreach` in each case should not produce a warning since the collection cannot be null after the first dereference.
```c#
[module: System.Runtime.CompilerServices.NonNullTypes]
class Enumerable
{
    public System.Collections.IEnumerator GetEnumerator() => throw null;
}
class Program
{
    static void F1(object[]? x)
    {
        foreach (var i in x) { } // warning: possibly null
        foreach (var i in x) { } // ok
    }
    static void F2(Enumerable? y)
    {
        foreach (var i in y) { } // warning: possibly null
        foreach (var i in y) { } // ok
    }
}
```

**Comment by cston** [comment]
See `NullableReferenceTypesTests.NotNullAfterDereference_ForEach`.

**Comment by jcouv** [comment]
It looks like I incorrectly adjusted `NotNullAfterDereference_ForEach`, so it's no longer verifying what we want. I'll rectify the test in my next PR.
